### ISSUE TITLE
SortedSet.collect cannot infer partial function

### ISSUE THREAD

**Issue by MasseGuillaume** [scala/bug#11055]
For this specific case:

```scala
collection.immutable.SortedSet(1).collect {
  val x = 1
  {
    case y if y > 1 => y
  }
}
```

compiles fine in 2.12 but in 2.13 I get the following error:

```
[error] The argument types of an anonymous function must be fully known. (SLS 8.5)
[error] Expected type was: ?
[error]       {
[error]       ^
```

**Comment by xuwei-k** [comment]
simpler example

```scala
package example

class A[X] {
  def collect[Y](pf: PartialFunction[X, Y]): Unit = ???
}

class B[X] extends A[X] {
  def collect[Y: Ordering](pf: PartialFunction[X, Y]): Unit = ???
}

object C {
  val b = new B[Int]
  b.collect{ case x if x > 2 => x } // success
  b.collect{ val z = 3 ; { case x if x > 2 => x } } // fail
}
```


`SortedSet` have two overload `collect`

https://static.javadoc.io/org.scala-lang/scala-library/2.13.0-M4/scala/collection/immutable/SortedSet.html

<img width="891" alt="screen shot 2018-08-07 at 16 56 33" src="https://user-images.githubusercontent.com/389787/43762425-e17f2966-9a62-11e8-83e0-46600ace1fb6.png">

**Comment by joshlemer** [comment]
@julienrf any ideas what to do here? maybe the SortedSet#collect should be renamed `collectSorted` or something?

**Comment by lrytz** [comment]
@adriaanm can inference be improved here? see @xuwei-k's [simpler example](https://github.com/scala/bug/issues/11055#issuecomment-410968847).

**Comment by adriaanm** [comment]
fixed by scala/scala#6871
